### DISCUSSION TITLE
python-oracledb is the Python module for Oracle Database

### DISCUSSION THREAD

**Discussion by cjbj** [oracle/python-oracledb#2] [Announcements]
Welcome to python-oracledb, the new, major [cx_Oracle driver](https://github.com/oracle/python-cx_Oracle) release. It comes with a brand new name **python-oracledb**, hence this new GitHub repository.

See the original [release announcement](https://cjones-oracle.medium.com/open-source-python-thin-driver-for-oracle-database-e82aac7ecf5a) for details.  Since that first release we have had several others, check the [release notes](https://python-oracledb.readthedocs.io/en/latest/release_notes.html).

## Interesting Stuff

- Python-oracledb is the new name for our popular Python cx_Oracle driver for Oracle Database.

- Python-oracledb 1.0 was a new major release - the successor to cx_Oracle 8.3.

- Python-oracledb is simple and small to install — under 15 MB (including its Python package dependencies): [`pip install oracledb`](https://python-oracledb.readthedocs.io/en/latest/user_guide/installation.html)

- Python-oracledb is now a Thin driver by default - it connects directly to Oracle Database without always needing Oracle Client libraries.

-  Python-oracledb runs on many platforms including favorites like Linux, macOS and Windows. It can also be used on platforms where Oracle Client libraries are not available (such as Apple M1, Alpine Linux, or IoT devices), or where the client libraries are not easily installed (such as some cloud environments).

- Python-oracledb has comprehensive functionality conforming to the Python Database API v2.0 Specification, with many additions and just a couple of exclusions.

- A “Thick” mode can be optionally enabled by an [application call](https://python-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#enabling-python-oracledb-thick-mode). This mode has similar functionality to cx_Oracle and supports Oracle Database features that extend the Python DB API. To use this mode, the widely used and tested Oracle Client libraries such as from Oracle Instant Client must be installed separately.

## Resources

Home page: [oracle.github.io/python-oracledb](https://oracle.github.io/python-oracledb/index.html)

Installation instructions: [python-oracledb.readthedocs.io/en/latest/installation.html](https://python-oracledb.readthedocs.io/en/latest/user_guide/installation.html)

Documentation: [python-oracledb.readthedocs.io/en/latest/index.html](https://python-oracledb.readthedocs.io/en/latest/index.html)

**Comment by tkofford** [comment]
Awesome, does this new version support Kerberos authentication?

**Reply by anthony-tuininga** [reply]
The thick mode does, not the thin mode, though, at least not yet!

**Reply by cjbj** [reply]
Kerberos support was mentioned by our VP last week as being important, so it's on our radar.

**Comment by fenchu** [comment]
Great, eager to test, Does this mean I do not oracle instant client anymore. It will mean 300MB less container size.

**Reply by cjbj** [reply]
@fenchu If the Thin mode has the features you want, then yes. If not, then you have to use Thick mode with Oracle Client libraries.  Check the doc: [Differences between python-oracledb Thin and Thick Modes](https://python-oracledb.readthedocs.io/en/latest/user_guide/appendix_b.html)

If Thin mode doesn't do what you need, please let us know so we can prioritize our projects.

**Reply by doerwalter** [reply]
It seems that when fetching date/time values from the database something has changed between `cx_Oracle` and `oracledb`:

```
Python 3.10.4 (main, Apr 26 2022, 19:42:59) [Clang 13.1.6 (clang-1316.0.21.2)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.3.0 -- An enhanced Interactive Python. Type '?' for help.
🐍 — 1 ❯ import oracledb, cx_Oracle
🐍 — 2 ❯ odb = oracledb.connect("user/pwd@db")
🐍 — 3 ❯ cdb = cx_Oracle.connect("user/pwd@db")
🐍 — 4 ❯ oc = odb.cursor()
🐍 — 5 ❯ cc = cdb.cursor()
🐍 — 6 ❯ cc.execute("select start_date from dba_scheduler_jobs where job_name = 'LOG_CLEANUP_JOB'").fetchone()
🐍 — 6 ❮ (datetime.datetime(2019, 3, 26, 4, 33),)
🐍 — 7 ❯ oc.execute("select start_date from dba_scheduler_jobs where job_name = 'LOG_CLEANUP_JOB'").fetchone()
🐍 — 7 ❮ (datetime.datetime(2019, 3, 26, 2, 33),)
🐍 — 8 ❯
```

Since both values are naive `datetime` objects one of the values must be wrong. The two hour difference is the offset between german local time and UTC.

When I do a `select start_date from dba_scheduler_jobs where job_name = 'LOG_CLEANUP_JOB'` in PL/SQL-Developer I get `26.03.19 04:33:00,000000 +02:00`

**Reply by doerwalter** [reply]
Also it seems that custom types are no longer supported. With

```sql
create or replace type varchars as table of varchar2(4000);
```

I get:

```
Python 3.10.4 (main, Apr 26 2022, 19:42:59) [Clang 13.1.6 (clang-1316.0.21.2)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.3.0 -- An enhanced Interactive Python. Type '?' for help.
🐍 — 1 — 18:20:16.308616 ❯ import oracledb, cx_Oracle
🐍 — 2 — 18:20:21.518450 ❯ odb = oracledb.connect("user/pwd@db")
🐍 — 3 — 18:20:26.352752 ❯ cdb = cx_Oracle.connect("user/pwd@db")
🐍 — 4 — 18:20:30.198178 ❯ oc = odb.cursor()
🐍 — 5 — 18:20:33.284193 ❯ cc = cdb.cursor()
🐍 — 8 — 18:20:52.522373 ❯ cc.execute("select varchars('foo', 'bar') from dual").fetchone()
🐍 — 8 — 18:34:14.630333 ❮ (<cx_Oracle.Object LL.VARCHARS at 0x10fadc370>,)
🐍 — 9 — 18:34:14.636108 ❯ oc.execute("select varchars('foo', 'bar') from dual").fetchone()
---------------------------------------------------------------------------
NotSupportedError                         Traceback (most recent call last)
Input In [9], in <cell line: 1>()
----> 1 oc.execute("select varchars('foo', 'bar') from dual").fetchone()

File ~/pyvenvs/default/lib/python3.10/site-packages/oracledb/cursor.py:378, in Cursor.execute(self, statement, parameters, **keyword_parameters)
    376 if parameters is not None:
    377     impl.bind_one(self, parameters)
--> 378 impl.execute(self)
    379 if impl.fetch_vars is not None:
    380     return self

File src/oracledb/impl/thin/cursor.pyx:122, in oracledb.thin_impl.ThinCursorImpl.execute()

File src/oracledb/impl/thin/protocol.pyx:294, in oracledb.thin_impl.Protocol._process_single_message()

File src/oracledb/impl/thin/protocol.pyx:295, in oracledb.thin_impl.Protocol._process_single_message()

File src/oracledb/impl/thin/protocol.pyx:255, in oracledb.thin_impl.Protocol._process_message()

File src/oracledb/impl/thin/protocol.pyx:234, in oracledb.thin_impl.Protocol._process_message()

File src/oracledb/impl/thin/messages.pyx:279, in oracledb.thin_impl.Message.process()

File src/oracledb/impl/thin/messages.pyx:785, in oracledb.thin_impl.MessageWithData._process_message()

File src/oracledb/impl/thin/messages.pyx:865, in oracledb.thin_impl.MessageWithData._process_row_data()

File src/oracledb/impl/thin/messages.pyx:591, in oracledb.thin_impl.MessageWithData._process_column_data()

File ~/pyvenvs/default/lib/python3.10/site-packages/oracledb/errors.py:103, in _raise_err(error_num, context_error_message, cause, **args)
    101     message = f"{message}\n{context_error_message}"
    102 exc_type = ERR_EXCEPTION_TYPES[error_num // 1000]
--> 103 raise exc_type(_Error(message)) from cause

NotSupportedError: DPY-3007: database type "DB_TYPE_OBJECT" is not supported
```

**Reply by anthony-tuininga** [reply]
Custom (Oracle Database) types are not supported with thin mode currently. That is on the list of things to implement. :-) If you use thick mode (by calling init_oracle_client() before creating your connection) it will work as expected.

**Reply by anthony-tuininga** [reply]
A similar issue may be occurring with your date scenario. If you call "alter session set time_zone" does that make a difference? If not, please open a new discussion or issue so we can see if the scenario can be replicated. Thanks!

**Reply by doerwalter** [reply]
OK, done for the timestamp issue: https://github.com/oracle/python-oracledb/issues/7

**Reply by doerwalter** [reply]
Yes, using `init_oracle_client()` makes the `varchars` example work:

```
Python 3.10.4 (main, Apr 26 2022, 19:42:59) [Clang 13.1.6 (clang-1316.0.21.2)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.3.0 -- An enhanced Interactive Python. Type '?' for help.
🐍 — 1 ❯ import oracledb
🐍 — 2 ❯ oracledb.init_oracle_client()
🐍 — 3 ❯ odb = oracledb.connect("user/pwd@db")
🐍 — 5 ❯ oc = odb.cursor()
🐍 — 7 ❯ oc.execute("select varchars('foo', 'bar') from dual").fetchone()
🐍 — 7 ❮ (<oracledb.Object LL.VARCHARS at 0x10bf06200>,)
🐍 — 8 ❯
```

**Comment by bburbidge** [comment]
FYI for anyone looking and working with SQLAlchemy, this is what needs to be done until the next SQLAlchemy release

`import sys`
`import oracledb`
`oracledb.version = "8.3.0"`
`sys.modules["cx_Oracle"] = oracledb`
`import cx_Oracle`

**Reply by cjbj** [reply]
Thanks - that's a good call to share this useful info.  Some related resources: 
- the user doc [Python Frameworks, SQL Generators, and ORMs](https://python-oracledb.readthedocs.io/en/latest/user_guide/appendix_c.html#python-frameworks-sql-generators-and-orms)
- blog post [Using python-oracledb 1.0 with SQLAlchemy, Pandas, Django and Flask](https://levelup.gitconnected.com/using-python-oracledb-1-0-with-sqlalchemy-pandas-django-and-flask-5d84e910cb19)
- blog post [Using SQLAlchemy 2.0 (development) with python-oracledb](https://medium.com/oracledevs/using-the-development-branch-of-sqlalchemy-2-0-with-python-oracledb-d6e89090899c)
